Fix finding ACID delete delta after minor compaction

harmandeeps opened the pull request:
Handling different formats of Delete delta file:
1. delete_delta_minWriteId_maxWriteID_statementId: default case
2. delete_delta_minWriteId_maxWriteID: After minor compaction, but we try to locate the delete delta file created: delete_delta_minWriteId_maxWriteID_0, which is not present. Hence, the deleted row is not removed from query results.

findepi commented on presto-hive/src/test/java/io/prestosql/plugin/hive/orc/TestOrcDeletedRows.java:
```diff
@@ -106,9 +106,40 @@ public void testDeletedLocationsOriginalFiles()
     {
-        Path deleteDeltaPath = new Path(path, AcidUtils.deleteDeltaSubdir(minWriteId, maxWriteId, statementId));
+        Path deleteDeltaPath;
+        if (statementId == -1) {
```
Can we use Optional<Integer> instead?

harmandeeps replied:
added.

findepi commented on presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveTransactionalTable.java:
```diff
@@ -123,6 +123,11 @@ private void doTestReadFullAcid(boolean isPartitioned, BucketingType bucketingTy
             // test filtering
             assertThat(query("SELECT col, fcol FROM " + tableName + " WHERE fcol = 1 ORDER BY col")).containsOnly(row(21, 1));
 
+            onHive().executeQuery(
+                    "INSERT INTO TABLE " + tableName + hivePartitionString + " VALUES (24, 4)");
+            onHive().executeQuery(
+                    "DELETE FROM " + tableName + " where fcol=4");
```
wrap each on one line

harmandeeps replied:
done.

findepi commented on presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveTransactionalTable.java:
```diff
@@ -123,6 +123,11 @@ private void doTestReadFullAcid(boolean isPartitioned, BucketingType bucketingTy
+                    "INSERT INTO TABLE " + tableName + hivePartitionString + " VALUES (24, 4)");
+            onHive().executeQuery(
+                    "DELETE FROM " + tableName + " where fcol=4");
+
```
add an additional verification after that, and before subsequent insert:
```
assertThat(query("SELECT col, fcol FROM " + tableName + " WHERE fcol = 1 ORDER BY col")).containsOnly(row(21, 1));
```

harmandeeps replied:
added.

findepi commented on presto-hive/src/main/java/io/prestosql/plugin/hive/AcidInfo.java:
```diff
@@ -272,7 +273,15 @@ public Builder addDeleteDelta(Path deleteDeltaPath, long minWriteId, long maxWri
+             // Handling different formats of Delete delta file:
+             // 1. delete_delta_minWriteId_maxWriteID_statementId: default case
+             // 2. delete_delta_minWriteId_maxWriteID: after minor compaction
+            if (deleteDeltaPath.getName().split("_").length == 5) {
```
Why are we parsing the `deleteDeltaPath` if `statementId` is given as a parameter here?

harmandeeps replied:
We fetch this ID from AcidUtils which has the following logic:
```
public int getStatementId() {
            return statementId == -1 ? 0 : statementId;
 }
```
So, we get 0 even if it is -1, that's why we need to figure it out from delete delta file name.

losipiuk replied:
I would prefer to create a shim over `AcidUtils.getAcidState` and do the postprocessing immediatelly after calling original method. So logic is not scattered around. WDYT?

harmandeeps commented:
> LGTM
> Minor comments
> 
> Please update commit and PR title to something like
> 
> ```
> Fix finding delete delta for data after minor compaction
> ```

Thank you for the review, I have addressed the comments, Please take a look.

findepi commented:
The build has multiple errors in presto-hive module. Can be unrelated, but would be great to have a clean build.

Can you please rebase to have CI rerun?
When doing so, can you please squash commits and change the commit title to 
```
Fix finding ACID delete delta after minor compaction
```
?

harmandeeps commented:
> The build has multiple errors in presto-hive module. Can be unrelated, but would be great to have a clean build.
> 
> Can you please rebase to have CI rerun?
> When doing so, can you please squash commits and change the commit title to
> 
> ```
> Fix finding ACID delete delta after minor compaction
> ```
> 
> ?

@findepi : I have squashed the commits, rebased. The build has passed. Please take a look.

findepi commented on presto-hive/src/main/java/io/prestosql/plugin/hive/AcidInfo.java:
```diff
@@ -114,13 +115,13 @@ public String toString()
                 @JsonProperty("minWriteId") long minWriteId,
                 @JsonProperty("maxWriteId") long maxWriteId,
-                @JsonProperty("statementId") int statementId)
+                @JsonProperty("statementId") OptionalInt statementId)
```
Use:
```
this. statementId  = requireNonNull(statementId, "statementId is null");
```

below

harmandeeps replied:
fixed.

findepi commented on presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java:
```diff
@@ -476,7 +476,15 @@ private void invokeNoMoreSplitsIfNecessary()
+                    // Handling different formats of Delete delta file:
+                    // 1. delete_delta_minWriteId_maxWriteID_statementId: default case
+                    // 2. delete_delta_minWriteId_maxWriteID: after minor compaction
+                    if (delta.getPath().getName().split("_").length == 5) {
```
I guess this cannot be as simple as
```
if (delta.getStatementId() != 0) {
```

because  statement id 0 seems  valid. Is it?

If we cannot get the information out of `delta` object normally,
i would rather use reflection than duplicate the logic from
https://github.com/apache/hive/blob/468907eab36f78df3e14a24005153c9a23d62555/ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java#L937-L959

findepi replied:
I've also opened https://issues.apache.org/jira/browse/HIVE-24082

harmandeeps replied:
Yeah, I was not able to use `getStatementId()` directly, I have changed the logic to compare the generated path from min, max, statement ID. Please let me know if it looks fine.

Thank you for filing the Hive issue.

findepi commented on presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcDeletedRows.java:
```diff
@@ -243,10 +243,19 @@ private RowId getRowId(int position)
-                deleteDeltaInfo.getMaxWriteId(),
-                deleteDeltaInfo.getStatementId()));
+        String deleteDeltaSubdir;
+        if (deleteDeltaInfo.getStatementId().isPresent()) {
+            deleteDeltaSubdir = deleteDeltaSubdir(
+                    deleteDeltaInfo.getMinWriteId(),
+                    deleteDeltaInfo.getMaxWriteId(),
+                    deleteDeltaInfo.getStatementId().getAsInt());
+        }
+        else {
+            deleteDeltaSubdir = deleteDeltaSubdir(
+                    deleteDeltaInfo.getMinWriteId(),
+                    deleteDeltaInfo.getMaxWriteId());
+        }
```
No need to wrap lines, they are not too long. This will also make the difference between `if` branches more apparent

```
if (deleteDeltaInfo.getStatementId().isPresent()) {
        deleteDeltaSubdir = deleteDeltaSubdir(deleteDeltaInfo.getMinWriteId(), deleteDeltaInfo.getMaxWriteId(), deleteDeltaInfo.getStatementId().getAsInt());
    }
    else {
        deleteDeltaSubdir = deleteDeltaSubdir(deleteDeltaInfo.getMinWriteId(), deleteDeltaInfo.getMaxWriteId());
    }
```

harmandeeps replied:
fixed.

findepi commented:
Merged as 88d45e7ce742117ef220ed7744b9628b54b56e84, thanks!
